### PR TITLE
Create wai-aria/role/contextual-roles.html

### PR DESCRIPTION
**Pull request by rahimabdi**
Closes https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/20

### CODE REVIEW

File: wai-aria/role/contextual-roles.html [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+- See Computed Role for more details on role computation when an element lacks required context: 
+https://w3c.github.io/core-aam/#roleMappingComputedRole 
+
+- See ARIA 1.2 spec for full listing of "Required Context Role" for each element:
+https://www.w3.org/TR/wai-aria-1.2/
+
+Identified roles with "Required Context" excluding abstract roles (e.g., child -> parent):
+
+- caption -> {figure, grid, table, treegrid}
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:27]: I don't recall that we test `figure -> caption` anywhere. That seems like a good one for this file.

Comment (cookiecrook) [reply wai-aria/role/contextual-roles.html:27]: The others may be in the grid-, table-, and tree-specific files, but maybe more variants are needed.

File: wai-aria/role/contextual-roles.html [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+- cell -> row
+- columnheader -> row
+- gridcell -> row
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:30]: At least some are in ./table-roles.html

File: wai-aria/role/contextual-roles.html [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+- listitem -> {list, directory} *Note: directory role is deprecated as of ARIA 1.2
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:31]: ./list-roles.html

File: wai-aria/role/contextual-roles.html [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+- menuitem -> {group, menu, menubar}
+- menuitemcheckbox -> {group, menu, menubar}
+- menuitemradio -> {group, menu, menubar}
+- option -> {group, listbox}
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:35]: ./menu-roles.html

File: wai-aria/role/contextual-roles.html [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+- row -> {grid, rowgroup, table, treegrid}
+- rowgroup -> {grid, table, treegrid}
+- rowheader -> row
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:38]: ./table-roles.html and/or ./grid-roles.html

File: wai-aria/role/contextual-roles.html [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+- tab -> tablist
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:39]: ./tab-roles.html

File: wai-aria/role/contextual-roles.html [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+- treeitem -> {group, tree}
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:40]: ./tree-roles.html

File: wai-aria/role/contextual-roles.html [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    2. If <footer> is scoped to <body>, it should be 'contentinfo' as expected
+
+    see: https://w3c.github.io/html-aam/#el-footer-ancestorbody -->
+<main>
+    <footer data-testname="footer scoped to main element" class="ex-generic">x</footer>
+</main>
+
+<main>
+    <div role="contentinfo" data-testname="contentinfo region scoped to main element" class="ex-generic">x</div>
+</main>
```
[the remark this answers is not in the document]
Comment (cookiecrook) [reply wai-aria/role/contextual-roles.html:91]: I can't think of another place where `footer`/`contentinfo` would or should be tested, so keeping them in this file seems good to me.

**Review by cookiecrook — CHANGES_REQUESTED** [review]
I'm sorry I didn't anticipate how many of these subtests would already be tested in other files. IIRC, I wrote the issue before those PRs were added.

You have lots of new tests here though, so any that aren't already tested could be moved to those related files, like tab-roles.html, list-roles.html, etc. I see at least a few that don't have another logical landing place, so keep those here?

File: wai-aria/role/contextual-roles.html (hunk 1 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:27]: @cookiecrook For your comment regarding `figure -> caption`, did you mean `caption -> figure`? I do have a test for a lone `role="caption"` within a `<p>` but want to ensure I didn't miss other related tests. Thank you!

File: wai-aria/role/contextual-roles.html [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+<div role="menuitemcheckbox" data-testname="orphan div with menuitemcheckbox role" class="ex-generic">x</div>
+<div role="menuitemradio" data-testname="orphan div with menuitemradio role" class="ex-generic">x</div>
+
+<!-- Testing orphaned group/listbox semantics -->
+<span role="option" data-testname="orphan span with option role" class="ex-generic">x</span>
+<button role="option" data-testname="orphan button with option role" data-expectedrole="button" class="ex">x</button>
+
+<!-- Testing orphaned tab semantics -->
+<button role="tab" data-testname="orphan button with tab role" data-expectedrole="button" class="ex">x</button>
+<span role="tab" data-testname="orphan span with tab role" class="ex-generic">x</span>
```
[the remark this answers is not in the document]
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html]: Resolved in 501c94759a6f689f41f44f5bcc6327be4335c1c9

File: wai-aria/role/contextual-roles.html [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+<!-- Testing orphaned tree semantics -->
+<button role="treeitem" data-testname="orphan button with treeitem role" data-expectedrole="button" class="ex">x</button>
+<span role="treeitem" data-testname="orphan span with treeitem role" class="ex-generic">x</span>
```
[the remark this answers is not in the document]
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html]: Resolved in 1d45c8119c10611d00d6216027d89d6dd8f0f920

File: wai-aria/role/contextual-roles.html [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+<!-- Testing orphaned grid semantics -->
+<button role="gridcell" data-testname="orphan button with gridcell role" data-expectedrole="button" class="ex">x</button>
+<span role="gridcell" data-testname="orphan span with gridcell role" class="ex-generic">x</span>
+
+<!-- Testing orphaned menu semantics -->
+<button role="menuitemcheckbox" data-testname="orphan button with menuitemcheckbox role" data-expectedrole="button" class="ex">x</button>
+<button role="menuitemradio" data-testname="orphan button with menuitemradio role" data-expectedrole="button" class="ex">x</button>
+<div role="menuitemcheckbox" data-testname="orphan div with menuitemcheckbox role" class="ex-generic">x</div>
+<div role="menuitemradio" data-testname="orphan div with menuitemradio role" class="ex-generic">x</div>
```
[the remark this answers is not in the document]
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html]: Added `menuitem`; resolved in f55021f9f148279c638074741cbe046d62589588

File: wai-aria/role/contextual-roles.html (hunk 2 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:30]: It looks like table-roles doesn't exist? Happy to create 👍🏾

File: wai-aria/role/contextual-roles.html (hunk 3 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:31]: Resolved as part of c511316c0ac4772351dc4f59afbcf89af7a6efe9

File: wai-aria/role/contextual-roles.html (hunk 4 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:35]: Done, please see c511316c0ac4772351dc4f59afbcf89af7a6efe9

File: wai-aria/role/contextual-roles.html (hunk 6 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:39]: Done; please see bf3c2f87f8a1b2459f8c30960a7df11468d7f666

File: wai-aria/role/contextual-roles.html (hunk 7 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:40]: Done; please see 1d45c8119c10611d00d6216027d89d6dd8f0f920

File: wai-aria/role/contextual-roles.html (hunk 5 above)
Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:38]: Done, please see aed8053a90073e291c93e45226bbe6fb2b5fb191

File: wai-aria/role/contextual-roles.html [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+</head>
+<body>
+
+<p>Verifies Required Accessibility Parent Roles from <a href="https://w3c.github.io/aria/#scope">5.2.7 Required Accessibility Parent Role</a></p>
+<p>Also verifies Allowed Accessibility Child Roles from <a href="https://w3c.github.io/aria/#mustContain">5.2.6 Allowed Accessibility Child Roles</a></p>
+
+<!-- Notes for "required context" testing:
+
+- See Computed Role for more details on role computation when an element lacks required context: 
+https://w3c.github.io/core-aam/#roleMappingComputedRole
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:20]: Cleaning up the whitespace lint to merge.
```suggestion
- See Computed Role for more details on role computation when an element lacks required context:
https://w3c.github.io/core-aam/#roleMappingComputedRole
```

File: wai-aria/role/contextual-roles.html (hunk 1 above)
Comment (cookiecrook) [reply wai-aria/role/contextual-roles.html:27]: I should have said [`figure -> figcaption`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figcaption) which I recalled (I was wrong) got a `caption` role according to HTML-AAM, [but it doesn't](https://w3c.github.io/html-aam/#el-figcaption), so no need to add another test here. 

Thread resolved.

File: wai-aria/role/contextual-roles.html (hunk 2 above)
Comment (cookiecrook) [reply wai-aria/role/contextual-roles.html:30]: Right... there's a `/html-aam/table-roles.html`, but I agree that this should be a new file in `/wai-aria`

Please add it, and include comment pointers to/from the related files:
- `/html-aam/table-roles.html`
- `/wai-aria/grid-roles.html`

such as:

```html
<!-- grid roles tested in ./grid-roles.html -->
```

and vice versa pointers back to your new file from the others. Thanks!

File: wai-aria/role/contextual-roles.html (hunk 4 above)
Comment (cookiecrook) [reply wai-aria/role/contextual-roles.html:35]: Sorry for the lack of clarity. I intended to comment on the 4 lines above, including the `menuitem` and variant roles. Those arent' the linked commit, but perhaps they landed elsewhere?

Comment (cookiecrook) [reply wai-aria/role/contextual-roles.html:35]: Looks like here: https://github.com/web-platform-tests/wpt/commit/bf3c2f87f8a1b2459f8c30960a7df11468d7f666 
Should those be in `./menu-roles.html`?

Comment (rahimabdi) [reply wai-aria/role/contextual-roles.html:35]: Yes! They got moved in a later commit: f55021f9f148279c638074741cbe046d62589588

**Review by rahimabdi — COMMENTED** [review]
Looks like some table-related work remains, will get on that...

File: wai-aria/role/table-roles.html [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    <script src="/resources/testdriver-actions.js"></script>
+    <script src="/wai-aria/scripts/aria-utils.js"></script>
+  </head>
+  <body>
+
+  <p>Tests <a href="https://w3c.github.io/aria/#table">table</a> and related roles.</p>
+
+<script>
+AriaUtils.verifyRolesBySelector(".ex");
+AriaUtils.verifyGenericRolesBySelector(".ex-generic");
```
Comment (cookiecrook) [inline wai-aria/role/table-roles.html:18]: Won't this throw an error since there aren't any tests?

Comment (cookiecrook) [reply wai-aria/role/table-roles.html:18]:
```
ERROR message: Selector passed in verifyRolesBySelector(".ex") should match at least one element.
```

https://wpt.fyi/results/wai-aria/role/table-roles.html?diff&filter=ADC&run_id=5131185556029440&run_id=4850203527479296

**Review by cookiecrook — CHANGES_REQUESTED** [review]

File: wai-aria/role/table-roles.html (hunk 13 above)
Comment (cookiecrook) [reply wai-aria/role/table-roles.html:18]: Both of these will error out (blocking further progress) since there aren't any subtests in the file.

**Review by rahimabdi — COMMENTED** [review]
@cookiecrook Ready for another review after creating wai-aria/role/table-roles.html and adding related documentation references for:
-  wai-aria/role/grid-roles.html
- wpt/html-aam/table.roles.html

File: wai-aria/role/contextual-roles.html [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    <!-- row -> ./grid-roles.html, ./table-roles.html  -->
+
+    <!-- rowgroup -> ./grid-roles.html, ./table-roles.html -->
+
+    <!-- rowheader -> ./grid-roles.html, ./table-roles.html -->
+
+    <!-- tab -> ./tab-roles.html -->
+
+    <!-- treeitem -> ./tree-roles.html -->
+
```
Comment (cookiecrook) [inline wai-aria/role/contextual-roles.html:69]:
```suggestion

```

File: wai-aria/role/table-roles.html [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    <title>Table Role Verification Tests</title>
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharnessreport.js"></script>
+    <script src="/resources/testdriver.js"></script>
+    <script src="/resources/testdriver-vendor.js"></script>
+    <script src="/resources/testdriver-actions.js"></script>
+    <script src="/wai-aria/scripts/aria-utils.js"></script>
+  </head>
+<body>
+
```
Comment (cookiecrook) [inline wai-aria/role/table-roles.html:13]:
```suggestion

```

File: wai-aria/role/table-roles.html [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+<body>
+    
+    <p>Tests <a href="https://w3c.github.io/aria/#table">table</a> and related roles.</p>
+
+<!-- HTML <table> Elements Testing  -->
+
+    <!-- <caption> -> wpt/html-aam/table-roles.html -->
+
+    <!-- <table> -> wpt/html-aam/table-roles.html -->
+
```
Comment (cookiecrook) [inline wai-aria/role/table-roles.html:21]:
```suggestion

```

File: wai-aria/role/table-roles.html [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+<!-- HTML <table> Elements Testing  -->
+
+    <!-- <caption> -> wpt/html-aam/table-roles.html -->
+
+    <!-- <table> -> wpt/html-aam/table-roles.html -->
+    
+    <!-- <tbody> -> wpt/html-aam/table-roles.html -->
+    
+    <!-- <td> -> wpt/html-aam/table-roles.html -->   
+
```
Comment (cookiecrook) [inline wai-aria/role/table-roles.html:25]:
```suggestion

    <!-- <td> -> wpt/html-aam/table-roles.html -->

```

File: wai-aria/role/table-roles.html [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+    <!-- <table> -> wpt/html-aam/table-roles.html -->
+    
+    <!-- <tbody> -> wpt/html-aam/table-roles.html -->
+    
+    <!-- <td> -> wpt/html-aam/table-roles.html -->   
+    
+    <!-- <tfoot> -> wpt/html-aam/table-roles.html -->
+
+    <!-- <th> -> wpt/html-aam/table-roles.html -->
+
```
Comment (cookiecrook) [inline wai-aria/role/table-roles.html:29]:
```suggestion

```

**Review by cookiecrook — COMMENTED** [review]
m. trailing whitespace lint.

**Review by cookiecrook — APPROVED** [review]

**Comment by cookiecrook** [comment]
As far as I can tell from digging through all the logs, this sink task is failing because ChromeDriver isn't downloading... It ran okay in Firefox and Safari, so why are the chrome-dev tests blocking the required sink-task check?

```
tools.wpt.browser.DownloadNotFoundError: No ChromeDriver download found to match browser version 120.0.6062.2 dev
```

**Comment by cookiecrook** [comment]
> We probably are overly eager to assume that a failure is caused by the PR and not by some browser vendor messing up their release process and it being impossible to download the browser/webdriver binary, but it's probably quite difficult to detect that specifically.

Thanks @gsnedders. So now that the need for the workaround is resolved, how should we re-trigger the tests? The re-run buttons seem to fail still.

**Comment by cookiecrook** [comment]
Hold off on more re-run attempts until after @gsnedders can respond to https://github.com/web-platform-tests/wpt/pull/42412#issuecomment-1765194111

**Comment by cookiecrook** [comment]
Attempting to re-run the full CI suite by closing and re-opening.

**Pull request merged**